syntax = "proto3";

package chat;

message LoginRequest {
	string name = 1;
	string auth = 2;
}

message LoginReply {
	enum statetype {
		ok = 0;
		auth_failed = 1;
		not_registered = 2;
		duplicate_login = 3;
		error = 4;
		banned = 5;
	}
	statetype state = 1;
	string token = 2;
}

message ChatSendRequest {

	message NormalMessage {
		string name = 1;
		string token = 2;
		string text = 3;
	}
	
	message RevokeMessage {
		string token = 1;
		enum revoketype {
			alter = 1;
			remove = 2;
		}
		revoketype type = 2;
		string target_uuid = 3;
		string text = 4;
	}
	
	oneof chat_message {
		NormalMessage normal_message = 1;
		RevokeMessage revoke_message = 2;
	}
	
}

message ChatSendReply {
	enum resulttype {
		ok = 0;
		bad_words = 1;
		error = 2;
	}
	resulttype result = 1;
}

message ChatMessage {
	string id = 1;
	string sender = 2;
	string text = 3;
	uint64 unix_time = 4;
}

message ChatRevokeMessage {
	string id = 1;
	string sender = 2;
	string text = 3;
	uint64 unix_time = 4;
	string target_id = 5;
	enum revoketype {
		alter = 1;
		remove = 2;
	}
	revoketype type = 6;
}

message NotifyChatMessageRequest {
	oneof req_message  {
		ChatMessage chat_message = 1;
		ChatRevokeMessage chat_revoke_message = 2;
	}
	
}

message NotifyChatMessageReply {
}

message GetLogRequest {
	string name = 1;
	string token = 2;
}

message GetLogReply {
	repeated ChatMessage chat_messages = 1;
}

message VerifyRequest {
	string name = 1;
	string token = 2;
}

message VerifyReply {
	bool ok = 1;
}


